### PR TITLE
Rust add list domains

### PR DESCRIPTION
**Pull request by benjamreis**

### CODE REVIEW

File: rust/src/xenctrl.rs [hunk 1]
```diff
@@ -117,9 +121,9 @@ impl Xenctrl {
       if ret != 1 || u32::from(info.domain) != dom_id {
         let error = self.get_last_error();
         if error.code == ErrorCode::None {
-          return Err(Error::new(ErrorCode::InvalidParam, ""));
+          return Err(Error::new(ErrorCode::InvalidParam, ""))
```
Comment (Wescoeur) [inline rust/src/xenctrl.rs:124]: Why this change?

File: rust/src/xenctrl.rs [hunk 2]
```diff
@@ -169,4 +173,39 @@ impl Xenctrl {
       }
     }
   }
+
+  pub fn list_domains (&self) -> Result<Vec<DomainInfo>> {
+    unsafe {
+      let max_doms: u32 = 1024;
+      let mut info: Vec<xenctrl_sys::xc_domaininfo_t> = Vec::with_capacity(max_doms as usize);
```
Comment (Wescoeur) [inline rust/src/xenctrl.rs]: Use `DomainInfo` instead of `xenctrl_sys::xc_domaininfo_t`  here.

File: rust/src/xenctrl.rs [hunk 3]
```diff
@@ -169,4 +173,39 @@ impl Xenctrl {
...
+      info.resize_with(max_doms as usize, Default::default);
+      let mut dom_id = 0;
+      let mut res: Vec<xenctrl_sys::xc_domaininfo_t> = Vec::new();
```
Comment (Wescoeur) [inline rust/src/xenctrl.rs]: Maybe rename `res` to `domains` to be more clear?

Comment (Wescoeur) [reply rust/src/xenctrl.rs]: Can you try without the type specifier? I suppose the inference algorithm can be smart here.

File: rust/src/xenctrl.rs [hunk 4]
```diff
@@ -169,4 +173,39 @@ impl Xenctrl {
...
+            let error = self.get_last_error();
+            if error.code == ErrorCode::None {
+              return Err(Error::new(ErrorCode::InvalidParam, ""))
+            } else {
+              return Err(error)
+            }
+          },
+          0 => break,
+          n => {
+            let info_res = &info[0..n as usize];
```
Comment (Wescoeur) [inline rust/src/xenctrl.rs]: You can use the suffix `usize` to avoid a cast.

File: rust/src/xenctrl.rs [hunk 5]
```diff
@@ -169,4 +173,39 @@ impl Xenctrl {
...
+            for i in 0..n {
+              let info_i = info[i as usize];
```
Comment (Wescoeur) [inline rust/src/xenctrl.rs]: Same idea here.

File: rust/src/xenctrl.rs [hunk 6]
```diff
@@ -196,7 +196,7 @@ impl Xenctrl {
           n => {
             let info_res = &info[0..n as usize];
             for i in 0..n {
-              let info_i = info[i as usize];
+              let info_i: DomainInfo = info[i as usize];
```
Comment (Wescoeur) [inline rust/src/xenctrl.rs]: Why?

File: rust/src/xenctrl.rs (hunk 1 above)
Comment (benjamreis) [reply rust/src/xenctrl.rs:124]: the `;`was useless, not necessary in this PR

File: rust/src/xenctrl.rs (hunk 6 above)
Comment (benjamreis) [reply rust/src/xenctrl.rs]: i was testing things, i already removed it

File: rust/src/xenctrl.rs (hunk 3 above)
Comment (benjamreis) [reply rust/src/xenctrl.rs]: indeed the inference works here! :)

**Review by Wescoeur — APPROVED** [review]
